nemo_aligner/models/nlp/gpt/gpt_steerlm_model.py

```diff
@@ -0,0 +1,357 @@
+                # while the fwd-bwd function is used to compute the loss of average microbatch loss
+                # i.e. L = sum_{b=0}^{num_batches} sum_{j=0}^{num_of_microbatches} [sum_{i=0}^{mbs} (L_i^{jb})] / (num_of_microbatches)
+                # to makie it equivalent to the loss we care, we need to scale loss by num_of_microbatches / num_batches
+                # i.e. L = sum_{b=0}^{num_batches} sum_{j=0}^{num_of_microbatches} [(num_of_microbatches)/num_batches * sum_{i=0}^{mbs} (L_i^{jb})] / (num_batches)
+                # also if we want to average the loss weighted by the inverse of number of loss_mass=1 tokens in the batch
+                batch["avg_num_valid_tokens_in_ub"][beg_idx:end_idx] = batch["loss_mask"][beg_idx:end_idx].sum(
+                    axis=-1
+                ).float().mean().repeat(response_size) / (num_of_microbatches / num_of_batches)
+                # compute the kl divergence between group['ws'] and group['base_weight']
+                # for numerical stability, add a small value to the base_weight_p
+                base_weight_p = base_weight_p + 1e-8
+                ws_weight = ws_weight + 1e-8
+                distance = torch.sum(ws_weight * torch.log(ws_weight / base_weight_p))
+                distances.append(distance)
+            # get average distance
+            distance = torch.stack(distances).mean()
+            distance = average_losses_across_data_parallel_group([distance])
+
+        set_sync_funcs(self, forward_only)
+        dp_size = int(parallel_state.get_data_parallel_world_size())
+        gbs = dp_size * gbs
+        mbs = int(self.cfg.data.steerlm2_micro_batch_size)
+        configure_batch_sizes(mbs=mbs, gbs=gbs, dp=dp_size)
+
+        # modify the iterator_k_split to iterate over all the data
+        data_iter = get_iterator_k_split(batch, get_num_microbatches())
+
+        fwd_loss_fn = self.get_forward_output_and_loss_func(forward_only)
+
+        losses_reduced = fwd_bwd_function(
+            forward_step_func=fwd_loss_fn,
+            data_iterator=data_iter,
+            model=self.model,
+            num_microbatches=get_num_microbatches(),
+            forward_only=forward_only,
+            micro_batch_size=get_micro_batch_size(),
+            seq_length=seq_length,
+        )
+        torch.cuda.synchronize()
+
+        # only the last stages of the pipeline return losses
+        if parallel_state.is_pipeline_last_stage():
+            loss_mean = torch.concat([loss["avg"] for loss in losses_reduced]).mean()
+        else:
+            loss_mean = torch.tensor(0.0).cuda()
+            distance = torch.tensor(0.0).cuda()
+        # Logging
+        torch.distributed.broadcast(loss_mean, get_last_rank())
+        torch.distributed.broadcast(distance, get_last_rank())
+        loss_value = loss_mean.detach().item()
+        metrics = {"loss": loss_value, "distance": distance.detach().item()}
```

Comment:
no need to detach if you're using .item() already

Reply:
done